k3s: structured netCfg

osnyx opened the pull request:
@flyingcircusio/release-managers

This makes several parts of our kubernetes network configuration more flexible by making the options of type "list". This simplifies overriding the network configuration for our IPv6-enabled cluster prototypes, and also fixes several places here these values are used.

Includes some additional resolver fixes left from #1920.

## Release process

- [x] Created changelog entry using `./changelog.sh`

## PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [x] if ticket is more urgent than within the next few days, directly contact a member of the Platform team
<!---->
- [x] set urgency and risk labels
- [ ] ensure the merge bot has determined a merge date
- [ ] ensure all checks are green
- [ ] get a review from a colleague

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [ ] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - make structure of our exposed k3s config options reflect the underlying kubelet options
  - structured options can be easier covered with individual checks for each configured option value
- [x] Security requirements tested? (EVIDENCE)
a dual-stack enabled cluster and inspect)
  - [x] in a dual-stack enabled cluster: inspect the generated `resolv.conf`s and the pods and service ipsgenerated.
  - [x] verify that name resolution from cluster frontend to pods still works
  - [x] inspected diff in k3s.service ExecStart command flags
  - [x] automated tests still pass

osnyx committed:
flyingcircus.kubernetes: make clusterDns, podCidr, serviceCidr a list

The underlying kubelet configuration supports multiple addresses per
option [1].
Some deployments have already utilised this for dual-stack clusters.
Changing the option type makes the processing throughout the config more
clean.

[1]: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/

ctheune commented on changelog.d/20251118_191955_PL-133889-clusterdns-firewall_scriv.md:
```diff
@@ -0,0 +1,39 @@
+     the main changelog (see below) is for.
+     -->
+
+- k3s clusters: options `clusterDns`, `podCidr`, `serviceCidr` are now a list
```
please put the emphasis of the semantics of the change into the changelog part. the impact should be very quickly understandable what the IMPACT is. So basically a single sentence:

k3s clusters with custom clusterDNS, podCidr, serviceCidr will fail to evaluate until adapted. See the change description below for details.

osnyx committed:
services/k3s/frontend: Do not include cluster DNS into local resolvers

Follow-up to https://github.com/flyingcircusio/fc-nixos/pull/1920, where
we did the same for all other k3s cluster node types. Quoting from
there:

> The hosts' resolv.conf is passed to the k3s cluster coredns, which
leads to coredns trying to sometimes resolve requests to itself. This
leads to timeouts and otherwise failing DNS request.

The frontend node runs not only a webgateway, but also a k3s.service.
The clusterDns resolvers continue to be written explicitly into the
haproxy config, so resolving should still work.

osnyx commented:
I'll shrink this down to the bare minimum for an actual clusterDns fix, the general restructuring of addresses will be done in #1970 then.

ctheune commented:
Say what? The actual fix has long been done. Big IMHO: this is a simple refactoring and should be wrapped up. I currently do not expect #1970 to land any time soon.

osnyx committed:
services/k3s: simplification of amending k3sFlags

`services.k3s.extraFlags` is of type "string or list of string".
We currently have some deplyoments that jsut want to amend some flags,
but due to the string nature they had to override and replicate *all*
flags. Turning this into a list allows just adding additional flags
while keeping existing ones.

osnyx committed:
services/k3s: expose nodeIps as an (internal) option

Currently used by some custom IPv6-enabled cluster deployments. Can be
cleaned up properly in PL-133774.

ctheune commented on nixos/services/k3s/frontend.nix:
```diff
@@ -319,7 +313,7 @@ in
         serverAddr = "https://${serverAddress}:6443";
         inherit tokenFile;
-        extraFlags = lib.concatStringsSep " " k3sFlags;
+        extraFlags = k3sFlags;
```
I'm a bit surprised by this. Why does this work? Is this hidden in the upstream code?

osnyx replied:
The upstream option takes either a `str` (and then can not merge additional strings into that) or a list of strings that can then be automerged.

dpausp reviewed (approved):
LGTM, we resolved the merge conflict together

leona-ya commented on nixos/roles/k3s/server.nix:
```diff
@@ -437,9 +435,18 @@ in
         flyingcircus.services.sensu-client =
           let
             kc = "${pkgs.k3s}/bin/k3s kubectl";
+
```
This breaks on k3s-server (probably not k3s-single-node), as the IP isn't available on the host.
